crates/virtio-queue/src/descriptor_utils.rs

```diff
@@ -0,0 +1,939 @@
+        descriptor_array_addr: GuestAddress,
+        mut buffers_start_addr: GuestAddress,
+        descriptors: Vec<(DescriptorType, u32)>,
+        spaces_between_regions: u32,
+    ) -> Result<DescriptorChain<&GuestMemoryMmap>> {
+        let descriptors_len = descriptors.len();
+        for (index, (type_, size)) in descriptors.into_iter().enumerate() {
+            let mut flags = 0;
+            if let DescriptorType::Writable = type_ {
+                flags |= VIRTQ_DESC_F_WRITE;
+            }
+            if index + 1 < descriptors_len {
+                flags |= VIRTQ_DESC_F_NEXT;
+            }
+
+            let index = index as u16;
+            let desc = virtq_desc {
+                addr: buffers_start_addr.raw_value().into(),
+                len: size.into(),
+                flags: flags.into(),
+                next: (index + 1).into(),
+            };
+
+            let offset = size + spaces_between_regions;
+            buffers_start_addr = buffers_start_addr
+                .checked_add(u64::from(offset))
+                .ok_or(Error::InvalidChain)?;
+
+            let _ = memory.write_obj(
+                desc,
+                descriptor_array_addr
+                    .checked_add(u64::from(index) * std::mem::size_of::<virtq_desc>() as u64)
+                    .ok_or(Error::InvalidChain)?,
+            );
+        }
+
+        let avail_ring = descriptor_array_addr
+            .checked_add(
+                u64::from(descriptors_len as u16) * std::mem::size_of::<virtq_desc>() as u64,
+            )
+            .ok_or(Error::InvalidChain)?;
+        let avail = virtq_avail {
+            flags: 0.into(),
+            idx: 1.into(),
+            ring: 0.into(),
+        };
+        let _ = memory.write_obj(avail, avail_ring);
+
+        let mut queue: Queue = Queue::new(MAX_QUEUE_SIZE).unwrap();
+        queue
+            .try_set_desc_table_address(descriptor_array_addr)
+            .unwrap();
+        queue.try_set_avail_ring_address(avail_ring).unwrap();
+        queue.set_ready(true);
+        let desc = queue.iter(memory).unwrap().next().unwrap();
+        Ok(desc.clone())
+    }
+
+    #[test]
+    fn reader_test_simple_chain() {
+        use DescriptorType::*;
+
+        let memory_start_addr = GuestAddress(0x0);
+        let memory = GuestMemoryMmap::from_ranges(&[(memory_start_addr, 0x10000)]).unwrap();
+
+        let chain = create_descriptor_chain(
+            &memory,
+            GuestAddress(0x0),
+            GuestAddress(0x100),
+            vec![
+                (Readable, 8),
+                (Readable, 16),
+                (Readable, 18),
+                (Readable, 64),
+            ],
+            0,
+        )
+        .expect("create_descriptor_chain failed");
+        let mut reader = Reader::new(&memory, chain).expect("failed to create Reader");
+        assert_eq!(reader.available_bytes(), 106);
+        assert_eq!(reader.bytes_read(), 0);
+
+        let mut buffer = [0 as u8; 64];
```

Comment:
nit: I do not think this case is required here. I think the read_exact should define this as byte just as well. Even if not, `0_u8` would probably be shorter.

Reply:
Do you mean that `writer_test_simple_chain()` test is not required?

Reply:
> I do not think this case is required here.

I think this was meant to mean: "cast" here... I was suggesting to drop the `as u8`. Sorry for the confusion...